apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/EventReportServiceClient.java

```diff
@@ -0,0 +1,183 @@
+                             );
+    }
+
+    @Override
+    public void boot() throws Throwable {
+
+    }
+
+    @Override
+    public void onComplete() throws Throwable {
+        reportStartingEvent();
+    }
+
+    @Override
+    public void shutdown() throws Throwable {
+        if (!CONNECTED.equals(status)) {
+            return;
+        }
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        final Event.Builder shutdownEvent = Event.newBuilder()
+                                                 .setUuid(UUID.randomUUID().toString())
+                                                 .setName("Shutdown")
+                                                 .setStartTime(System.currentTimeMillis())
+                                                 .setEndTime(System.currentTimeMillis())
+                                                 .setMessage("Shutting down Java Application")
+                                                 .setType(Type.Normal)
+                                                 .setSource(
+                                                     Source.newBuilder()
+                                                           .setService(Config.Agent.SERVICE_NAME)
+                                                           .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                                                           .build()
+                                                 );
+
+        final StreamObserver<Event> collector = eventServiceStub.collect(new StreamObserver<Commands>() {
+            @Override
+            public void onNext(final Commands commands) {
+                ServiceManager.INSTANCE.findService(CommandService.class).receiveCommand(commands);
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                LOGGER.error("Failed to report shutdown event.", t);
+                // Ignore status change at shutting down stage.
+                latch.countDown();
+            }
+
+            @Override
+            public void onCompleted() {
+                latch.countDown();
+            }
+        });
+
+        collector.onNext(shutdownEvent.build());
+        collector.onCompleted();
+        latch.await();
+    }
+
+    @Override
+    public void statusChanged(final GRPCChannelStatus status) {
+        this.status = status;
+
+        if (!CONNECTED.equals(status)) {
+            return;
+        }
+
+        final Channel channel = ServiceManager.INSTANCE.findService(GRPCChannelManager.class).getChannel();
+        eventServiceStub = EventServiceGrpc.newStub(channel);
+        eventServiceStub = eventServiceStub.withDeadlineAfter(GRPC_UPSTREAM_TIMEOUT, TimeUnit.SECONDS);
+
+        reportStartingEvent();
+    }
+
+    private void reportStartingEvent() {
+        if (reported.compareAndSet(false, true)) {
+            return;
+        }
+
+        startingEvent.setEndTime(System.currentTimeMillis());
```

Comment:
In `onError`, you will set `reported = false` again, then it could cause the end time of starting event will be reset in the next report. I think you don't want that.